File: Week1_Design Patterns and Principles/Ex_2_Implementing_the_Factory_Method_Pattern/Code/com/example/factory/documents/ExcelDocument.java
package com.example.factory.documents;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
public class ExcelDocument implements Document {
    private LocalDateTime creationTime;
    private String fileName; 
    public ExcelDocument() {
        this.creationTime = LocalDateTime.now();
        this.fileName = "Spreadsheet_" + creationTime.format(DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss")) + ".xlsx";
    }
    @Override
    public void open() {
        System.out.println("Opening Excel document: " + fileName);
        System.out.println("Loading Microsoft Excel application...");
    }
    @Override
    public void save() {
        System.out.println("Saving Excel document: " + fileName);
        System.out.println("Workbook saved in .xlsx format");
    }
    @Override
    public void close() {
        System.out.println("Closing Excel document: " + fileName);
        System.out.println("Microsoft Excel application closed");
    }
    @Override
    public void print() {
        System.out.println("Printing Excel document: " + fileName);
        System.out.println("Printing all worksheets...");
    }
    @Override
    public String getDocumentType() {
        return "Microsoft Excel Spreadsheet";
    }
    @Override
    public LocalDateTime getCreationTime() {
        return creationTime;
    }
}

File: Week1_Design Patterns and Principles/Ex_2_Implementing_the_Factory_Method_Pattern/Code/Documents/ExcelDocument.cs
using System;
using System.Globalization;

namespace FactoryDemo.Documents;

public sealed class ExcelDocument : IDocument
{
    private readonly DateTime _creationTime = DateTime.Now;
    private readonly string _fileName;

    public ExcelDocument()
    {
        _fileName = $"Spreadsheet_{_creationTime.ToString("yyyyMMdd_HHmmss", CultureInfo.InvariantCulture)}.xlsx";
    }

    public void Open()  => Console.WriteLine($"Opening Excel document: {_fileName}\nLoading Microsoft Excel application…");
    public void Save()  => Console.WriteLine($"Saving Excel document: {_fileName}\nWorkbook saved in .xlsx format");
    public void Close() => Console.WriteLine($"Closing Excel document: {_fileName}\nMicrosoft Excel application closed");
    public void Print() => Console.WriteLine($"Printing Excel document: {_fileName}\nPrinting all worksheets…");

    public string GetDocumentType()  => "Microsoft Excel Spreadsheet";
    public DateTime GetCreationTime() => _creationTime;
}
